### PR TITLE
sqlinstance: add `binary_version` column to instances table

### PR DESCRIPTION
**Pull request by healthy-pod**
This code change adds a `binary_version` column to the instances table.

This is done by adding the column to the bootstrap schema for system.sql_instances,
and piggy-backing on the existing code for the V23_1_SystemRbrReadNew migration
that overwrites the live schema for this table by the bootstrap copy.

This redefinition of the meaning of the V23_1_SystemRbrReadNew is backward-incompatible
and is only possible because this commit is merged before the v23.1 branch is cut.

Release note: None
Epic: [CRDB-20860](https://cockroachlabs.atlassian.net/browse/CRDB-20860)

### CODE REVIEW

**Review by knz — COMMENTED** [review]
it would be good to link your PR to one of our epics.

> Reviewed 23 of 23 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98830#-:-NQi_UgXB4KfxSodQQMJ:buaok1d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @healthy-pod and @JeffSwenson)

___
*[`pkg/sql/sqlinstance/instancestorage/row_codec.go` line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98830#-NQiZv0mAYd-P6NUj8Q2:-NQiZv0mAYd-P6NUj8Q3:b-xubz5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/99eb0be444c1e85453d1f4408a93fd7443c5a579/pkg/sql/sqlinstance/instancestorage/row_codec.go#L341)):*
> ```Go
> 		binaryVersionColumnIdx: func(datum tree.Datum) error {
> 			if datum != tree.DNull {
> 				binaryVersion = roachpb.MustParseVersion(string(tree.MustBeDString(datum)))
> ```

i recommend using the variant that checks the format and returns an error in case the value is invalid.

File: pkg/upgrade/upgrades/alter_sql_instances_binary_version_test.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+		[]string{},
+		validationSchemas,
+		true, /* expectExists */
+	)
+}
+
+// getDeprecatedSqlInstancesDescriptorWithoutBinaryVersion returns the system.sql_instances
+// table descriptor that was being used before adding a new column in the
+// current version.
+func getDeprecatedSqlInstancesDescriptorWithoutBinaryVersion() *descpb.TableDescriptor {
```
Comment (JeffSwenson) [inline pkg/upgrade/upgrades/alter_sql_instances_binary_version_test.go:106]: It's possible to bootstrap the DB with the V22_2 system database. It makes writing upgrade tests a lot easier and the tests are more representative of how the upgrades will run in production. See https://github.com/cockroachdb/cockroach/pull/98613 for an example.

File: pkg/upgrade/upgrades/alter_sql_instances_binary_version.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	tableID := catid.DescID(keys.SQLInstancesTableID)
+	for _, op := range []operation{
+		{
+			name:           "add-sql-instances-binary-version-column",
+			schemaList:     []string{"binary_version"},
+			query:          addBinaryVersionColumnToSQLInstancesTableStmt,
+			schemaExistsFn: columnExists,
+		},
+	} {
+		if err := migrateTable(ctx, cs, d, op, tableID, systemschema.SQLInstancesTable()); err != nil {
```
Comment (JeffSwenson) [inline pkg/upgrade/upgrades/alter_sql_instances_binary_version.go:41]: This upgrade isn't actually doing anything in production. The RBR index migration replaces the descriptor with the statically defined descriptor, which would include the column you are adding.

https://github.com/cockroachdb/cockroach/blob/85c6e3874ed6f5e5e47da278b06e26e09d78bfb6/pkg/upgrade/upgrades/system_rbr_indexes.go#L104

Comment (healthy-pod) [reply pkg/upgrade/upgrades/alter_sql_instances_binary_version.go:41]: Should I then remove the migration and keep the test case (run it after RBR index migration)? Anything else I should do? Actually I can't find the version associated with the RBR index migration. Can you please give more details?

**Comment by healthy-pod** [comment]
> it would be good to link your PR to one of our epics.

Done

> i recommend using the variant that checks the format and returns an error in case the value is invalid.

Done

File: pkg/upgrade/upgrades/alter_sql_instances_binary_version_test.go (hunk 1 above)
Comment (healthy-pod) [reply pkg/upgrade/upgrades/alter_sql_instances_binary_version_test.go]: Done

File: pkg/upgrade/upgrades/alter_sql_instances_binary_version.go (hunk 2 above)
Comment (JeffSwenson) [reply pkg/upgrade/upgrades/alter_sql_instances_binary_version.go:41]: This is the upgrade and version gate that will replace the descriptor with the new version. https://github.com/cockroachdb/cockroach/blob/ea389ffee029b69ed9f6c5957fd0997cf3f0145a/pkg/upgrade/upgrades/upgrades.go#L250

You should be able to delete your version key and upgrade logic. I think it is a good idea to keep the test. It documents the fact it is an intended side effect of the RBR migration.

**Comment by healthy-pod** [comment]
Ready for another look (pending clarification on @knz's question [here](https://cockroachlabs.slack.com/archives/C02HWA24541/p1679344939131769)).

**Review by knz — COMMENTED** [review]
If you remove the migration, then this entire PR doesn't do what it says on the label, as it does _not_ add a binary_version column to the instances table. I'll want the clarification before I even start the review.

> Reviewed 1 of 2 files at r2, 12 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98830#-:-NR-yFYr1mqsHGMN7RYM:b-g6kkij)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @healthy-pod and @JeffSwenson)

**Review by knz — APPROVED** [review]
Now I understand. Great work. :lgtm: modulo updating the commit message (and PR description)

(See below for another followup action)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98830#-:-NR00juQELCEAFqo6bBH:b-44e1we)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @healthy-pod and @JeffSwenson)

___
*[`-- commits` line 5 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98830#-NR0-m6CABk_Yixvr42K):*
Now that I understand what this is doing, _please_ add more words to this commit message to explain it as well:

"""
This is done by adding the column to the bootstrap schema for system.sql_instances,
and piggy-backing on the existing code for the `V23_1_SystemRbrReadNew` migration
that overwrites the live schema for this table by the bootstrap copy.

This redefinition of the meaning of the `V23_1_SystemRbrReadNew` is backward-incompatible
and is only possible because this commit is merged before the v23.1 branch is cut.
"""
___
*[`pkg/upgrade/upgrades/upgrades.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98830#-NR00R-p4W0ISKLXeqzZ:-NR00R-p4W0ISKLXeqz_:b-y1qdy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/317450575cedb1a186f5c2f548382981ef095434/pkg/upgrade/upgrades/upgrades.go#L254)):*
> ```Go
> 		toCV(clusterversion.V23_1_SystemRbrReadNew),
> 		upgrade.NoPrecondition,
> 		backfillRegionalByRowIndex,
> ```

Please file an issue to be processed at _very high priority_ by either the Serverless or Schema team, to enhance this function with a hardwired copy of the bootstrap descriptors at that version.

Otherwise, any additional changes to these tables in future CRDB versions will result in schema corruption for folk upgrading from an older version.

**Review by jeffswenson — APPROVED** [review]
LGTM

**Comment by healthy-pod** [comment]
TFTRs!

bors r=[knz,JeffSwenson]

**Comment by knz** [comment]
bors r=knz,JeffSwenson single on

**Pull request merged**